Logging inside of a `@match` throws an `AssertionError`

MRE:
```julia
julia> @match 1 begin
          1 => @info "aaa"
       end
ERROR: AssertionError: Assertion to tell the compiler about the definedness of this variable
Stacktrace:
 [1] top-level scope
   @ REPL[30]:2
 [2] macro expansion
   @ logging/logging.jl:415 [inlined]
```

```julia
julia> pkgversion(Match)
v"2.4.0"

julia> versioninfo()
Julia Version 1.12.0-rc1
Commit 228edd6610b (2025-07-12 20:11 UTC)
Build Info:
  Official https://julialang.org release
Platform Info:
  OS: macOS (arm64-apple-darwin24.0.0)
  CPU: 10 × Apple M1 Max
  WORD_SIZE: 64
  LLVM: libLLVM-18.1.7 (ORCJIT, apple-m1)
  GC: Built with stock GC
Threads: 1 default, 1 interactive, 1 GC (on 8 virtual cores)
```

--- a/LICENSE.md
+++ b/LICENSE.md
@@ -2,7 +2,7 @@
 
 The Match.jl package is licensed under the MIT Expat License:
 
-Copyright (c) 2013-2023: Neal Gafter, Kevin Squire, RelationalAI, Inc, and contributors.
+Copyright (c) 2013-2025: Neal Gafter, Kevin Squire, RelationalAI, Inc, and contributors.
 
 Permission is hereby granted, free of charge, to any person obtaining
 a copy of this software and associated documentation files (the

--- a/Project.toml
+++ b/Project.toml
@@ -1,6 +1,6 @@
 name = "Match"
 uuid = "7eb4fadd-790c-5f42-8a69-bfa0b872bfbf"
-version = "2.4.0"
+version = "2.4.1"
 authors = ["Neal Gafter <neal@gafter.com>", "Nate Nystrom <nate.nystrom@relational.ai>", "Kevin Squire <kevin.squire@gmail.com>"]
 
 [deps]

--- a/src/binding.jl
+++ b/src/binding.jl
@@ -167,16 +167,6 @@ function bind_pattern!(
         # a constant, e.g. a regular expression, version number, raw string, etc.
         pattern = BoundIsMatchTestPattern(input, BoundExpression(location, source), false)
 
-    elseif is_expr(source, :macrocall)
-        # We permit custom string macros as long as they do not contain any unbound
-        # variables.  We accomplish that simply by expanding the macro.  Macros that
-        # interpolate, like lazy"", will fail because they produce a `call` rather
-        # than an object.  Also, we permit users to define macros that expand to patterns.
-        while is_expr(source, :macrocall)
-            source = macroexpand(binder.mod, source; recursive = false)
-        end
-        (pattern, assigned) = bind_pattern!(location, source, input, binder, assigned)
-
     elseif is_expr(source, :$)
         # an interpolation
         interpolation = source.args[1]
@@ -673,13 +663,12 @@ function bind_case(
     case,
     predeclared_temps,
     binder::BinderContext)::BoundCase
+    case = macroexpand(binder.mod, case)
+
     while true
         # do some rewritings if needed
-        if is_expr(case, :macrocall)
-            # expand top-level macros only
-            case = macroexpand(binder.mod, case, recursive=false)
 
-        elseif is_expr(case, :tuple, 2) && is_case(case.args[2]) && is_expr(case.args[2].args[2], :if, 2)
+        if is_expr(case, :tuple, 2) && is_case(case.args[2]) && is_expr(case.args[2].args[2], :if, 2)
             # rewrite `pattern, if guard end => result`, which parses as
             # `pattern, (if guard end => result)`
             # to `(pattern, if guard end) => result`

--- a/src/match_return.jl
+++ b/src/match_return.jl
@@ -1,38 +1,34 @@
-"""
-    @match_fail
-
-This statement permits early-exit from the value of a @match case.
-The programmer may write the value as a `begin ... end` and then,
-within the value, the programmer may write
-
-    @match_fail
-
-to cause the case to terminate as if its pattern had failed.
-This permits cases to perform some computation before deciding if the
-rule "*really*" matched.
-"""
 macro match_fail()
-    # These are rewritten during expansion of the `@match` macro,
-    # so the actual macro should not be used directly.
-    error("$(__source__.file):$(__source__.line): @match_fail may only be used within the value of a @match case.")
+    # These are rewritten during expansion of the `@match` macro.
+    Expr(:call, :__match_fail__, QuoteNode(__source__))
 end
 
-"""
-    @match_return value
-
-This statement permits early-exit from the value of a @match case.
-The programmer may write the value as a `begin ... end` and then,
-within the value, the programmer may write
-
-    @match_return value
-
-to terminate the value expression **early** with success, with the
-given value.
-"""
 macro match_return(x)
-    # These are rewritten during expansion of the `@match` macro,
-    # so the actual macro should not be used.
-    error("$(__source__.file):$(__source__.line): @match_return may only be used within the value of a @match case.")
+    # These are rewritten during expansion of the `@match` macro.
+    Expr(:call, :__match_return__, QuoteNode(__source__), esc(x))
+end
+
+function __match_fail__(location)
+    error("$(location.file):$(location.line): @match_fail may only be used within the value of a @match case.")
+end
+
+function __match_return__(location, _)
+    error("$(location.file):$(location.line): @match_return may only be used within the value of a @match case.")
+end
+
+# is f a reference to the __match_fail__ function?
+function is_match_fail(f)
+    f == Match.__match_fail__ || f == GlobalRef(Match, :__match_fail__)
+end
+
+# is f a reference to the __match_return__ function?
+function is_match_return(f)
+    f == Match.__match_return__ || f == GlobalRef(Match, :__match_return__)
+end
+
+# is f an early exit function?
+function is_early_exit(f)
+    is_match_fail(f) || is_match_return(f)
 end
 
 #
@@ -78,30 +74,18 @@ end
 function adjust_case_for_return_macro(__module__, location, pattern, result, predeclared_temps)
     found_early_exit::Bool = false
 
-    # Check for the presence of early exit macros @match_return and @match_fail
+    # Check for the presence of early exit macros
     function adjust_top(p)
-        if is_macro(p, :var"@match_return", 3)
-            # :(@match_return e) -> :($value = $e; @goto $label)
-            found_early_exit = true
-            # expansion of the result will be done later by ExpressionRequiringAdjustmentForReturnMacro 
-            return p
-        elseif is_macro(p, :var"@match_fail", 2)
-            # :(@match_fail) -> :($value = $MatchFaulure; @goto $label)
-            found_early_exit = true
-            # expansion of the result will be done later by ExpressionRequiringAdjustmentForReturnMacro 
-            return p
-        elseif is_macro(p, :var"@match", 4)
-            # Nested uses of @match should be treated as independent
-            return macroexpand(__module__, p)
-        elseif is_expr(p, :macrocall)
-            # It is possible for a macro to expand into @match_fail, so only expand one step.
-            return adjust_top(macroexpand(__module__, p; recursive = false))
-        else
-            return p
+        if !found_early_exit && is_expr(p, :call)
+            f = p.args[1]
+            if is_early_exit(f)
+                found_early_exit = true
+            end
         end
+        return p
     end
+    MacroTools.postwalk(adjust_top, result)
 
-    rewritten_result = MacroTools.prewalk(adjust_top, result)
     if found_early_exit
         # Since we found an early exit, we need to predeclare the temp to ensure
         # it is in scope both for where it is written and in the constructed where clause.
@@ -109,19 +93,13 @@ function adjust_case_for_return_macro(__module__, location, pattern, result, pre
         push!(predeclared_temps, value_symbol)
         
         # Defer generation of the label and branch, so we get a unique label in case it needs to be generated more than once.
-        where_expr = where_expression_requiring_adjustment_for_return_macro(value_symbol, location, rewritten_result)
+        where_expr = where_expression_requiring_adjustment_for_return_macro(value_symbol, location, result)
         new_pattern = :($pattern where $where_expr)
         new_result = value_symbol
         (new_pattern, new_result)
     else
-        (pattern, rewritten_result)
+        (pattern, result)
     end
-end
-
-function is_macro(x, name::Symbol, arity::Int)
-    return is_expr(x, :macrocall) && ## length(x.args) == arity &&
-        (x.args[1] == name ||
-         x.args[1] == Expr(:., Symbol(string(@__MODULE__)), QuoteNode(name)))
 end
 
 const marker_for_where_expression_requiring_adjustment_for_return_macro =
@@ -147,15 +125,20 @@ function code_for_expression(x::Expr)
     label = gensym("early_label")
 
     function adjust_top(result)
-        if is_macro(result, :var"@match_return", 3)
+        is_expr(result, :call) || return result
+        f = result.args[1]
+        new_result = if is_match_return(f)
+            _, source, value = result.args
             # :(@match_return e) -> :($value = $e; @goto $label)
-            return Expr(:block, result.args[2], :($value_symbol = $(result.args[3])), :(@goto $label))
-        elseif is_macro(result, :var"@match_fail", 2)
+            Expr(:block, source, :($value_symbol = $(value)), :(@goto $label))
+        elseif is_match_fail(f)
+            _, source = result.args
             # :(@match_fail) -> :($value = $MatchFaulure; @goto $label)
-            return Expr(:block, result.args[2], :($value_symbol = $MatchFailure), :(@goto $label))
+            Expr(:block, source, :($value_symbol = $MatchFailure), :(@goto $label))
         else
-            return result 
+            result
         end
+        return new_result
     end
 
     rewritten_result = MacroTools.prewalk(adjust_top, expression_containing_macro)

--- a/src/match_return.jl
+++ b/src/match_return.jl
@@ -75,7 +75,7 @@ function adjust_case_for_return_macro(__module__, location, pattern, result, pre
     found_early_exit::Bool = false
 
     # Check for the presence of early exit macros
-    function adjust_top(p)
+    function check_for_early_exit(p)
         if !found_early_exit && is_expr(p, :call)
             f = p.args[1]
             if is_early_exit(f)
@@ -84,7 +84,7 @@ function adjust_case_for_return_macro(__module__, location, pattern, result, pre
         end
         return p
     end
-    MacroTools.postwalk(adjust_top, result)
+    MacroTools.postwalk(check_for_early_exit, result)
 
     if found_early_exit
         # Since we found an early exit, we need to predeclare the temp to ensure

--- a/test/Project.toml
+++ b/test/Project.toml
@@ -1,5 +1,5 @@
 [deps]
 Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
-ReTest = "e0db7c4e-2690-44b9-bad6-7687da720f89"
 Match = "7eb4fadd-790c-5f42-8a69-bfa0b872bfbf"
 MacroTools = "1914dd2f-81c6-5fcd-8719-6d5c9610ff09"
+Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"

--- a/test/match_return.jl
+++ b/test/match_return.jl
@@ -33,11 +33,9 @@ file = Symbol(@__FILE__)
     let line = 0
         try
             line = (@__LINE__) + 1
-            @eval @match_return 2
+            @match_return 2
             @test false
-        catch ex
-            @test ex isa LoadError
-            e = ex.error
+        catch e
             @test e isa ErrorException
             @test e.msg == "$file:$line: @match_return may only be used within the value of a @match case."
         end
@@ -48,11 +46,9 @@ end
     let line = 0
         try
             line = (@__LINE__) + 1
-            @eval @match_fail
+            @match_fail
             @test false
-        catch ex
-            @test ex isa LoadError
-            e = ex.error
+        catch e
             @test e isa ErrorException
             @test e.msg == "$file:$line: @match_fail may only be used within the value of a @match case."
         end

--- a/test/runtests.jl
+++ b/test/runtests.jl
@@ -2,7 +2,7 @@ module Rematch2Tests
 
 using Match
 using Match: topological_sort, @__match__
-using ReTest
+using Test
 using Random
 using MacroTools: MacroTools
 
@@ -15,7 +15,5 @@ include("topological.jl")
 include("match_return.jl")
 include("test_ismatch.jl")
 include("matchtests.jl")
-
-retest(Rematch2Tests)
 
 end # module

